classes/R4Report.php

```diff
@@ -0,0 +1,495 @@
+			else {
+				break;
+			}
+			$count++;
+		}
+		
+		foreach($studyQuestions as $study => $label) {
+			if($study == "") continue;
+			
+			foreach($surveyQuestions as $indexQuestion => $survey) {
+				$this->addTooltipCounts($study,$survey);
+			}
+		}
+	}
+	
+	public function createQuestion_2() {
+		$studyQuestions = ProjectData::getArrayStudyQuestion_2();
+		$surveyQuestions = ProjectData::getRowQuestionsResponseRate();
+		$surveyQuestions2 = ProjectData::getRowQuestionsParticipantPerception();
+		
+	}
+	
+	public function createQuestion_3() {
+		$studyQuestions = ProjectData::getArrayStudyQuestion_3();
+		$surveyQuestions = $this->getQuestion3SurveyFields();
+		
+		$custom_filters = $this->getProjectSetting('custom-filter',$this->projectId);
+		$count = 1;
+		foreach($custom_filters as $sstudy) {
+			if($count < 11 && $sstudy != "") {
+				$studyQuestions[$sstudy] = "Custom site value " . $count;
+			}
+			else {
+				break;
+			}
+			$count++;
+		}
+		
+		foreach($studyQuestions as $study => $label) {
+			foreach($surveyQuestions as $indexQuestion => $survey) {
+				$this->addTooltipCounts($study,$survey);
+			}
+		}
+	}
+	
+	public function getQuestion3SurveyFields() {
+		$surveyQuestions = ProjectData::getRowQuestions();
+		$surveyFields = [];
+		
+		for ($question = 2; $question < 5; $question++) {
+			$option = explode("-", $surveyQuestions[$question]);
+			for ($i = $option[0]; $i < $option[1]; $i++) {
+				$surveyFields[] = "rpps_s_q".$i;
+			}
+		}
+		
+		return $surveyFields;
+	}
+	public function createTableData() {
+		$allData_array = [];
+		$allDataTooptip_array = [];
+		foreach($this->surveyPercentages as $study => $studyDetails) {
+			foreach($studyDetails as $survey => $surveyDetails) {
+				if(array_key_exists($study,ProjectData::getArrayStudyQuestion_1())) {
+					$alldata_array["nofilter"][$study][$survey] = $surveyDetails;
+					$allDataTooptip_array["nofilter"][$study][$survey] = $this->tooltipTextArray[$study][$survey];
+				}
+				else {
+					$alldata_array[$study][$survey] = $surveyDetails;
+					$allDataTooptip_array[$study][$survey] = $this->tooltipTextArray[$study][$survey];
+				}
+			}
+			
+			foreach($this->surveyPercentagesInstitutions[$study] as $institution => $institutionDetails) {
+				foreach($institutionDetails as $survey => $surveyDetails) {
+					$surveyIndex = array_search($survey, ProjectData::getRowQuestionsParticipantPerception());
+					if($surveyIndex === false) {
+						$surveyIndex = array_search($survey, ProjectData::getRowQuestionsResponseRate());
+					}
+					$allData_array[$study]["institutions"][$surveyIndex][$institution] = $surveyDetails;
+				}
+			}
+		}
+		
+		$this->tableData = [
+			"data" => $allData_array,
+			"tooltip" => $allDataTooptip_array,
+			"legend" => true
+		];
+	}
+	
+	public function getEventId() {
+		if(!isset($this->eventId)) {
+			$this->eventId = $this->getFirstEventId();
+		}
+		return $this->eventId;
+	}
+	
+	public function applyFilterToData($filterLogic) {
```

Comment:
I really like this way of filtering so you get only the records that match. Way faster than looping by using REDCap's functions